rfcs/2024-08-21-motion-presets.md

```diff
@@ -0,0 +1,523 @@
+  </td>
+  <td width="500px">
+
+https://github.com/user-attachments/assets/2ec967a5-1e17-443a-a3a9-230962f6bd9a
+
+_Previews are just examples of presets. They don't use actual durations and easings yet_
+
+  </td>
+</tr>
+</table>
+
+<details>
+<summary>Alternate Morph APIs</summary>
+
+#### 2. Using `motion` from framer-motion
+
+```jsx
+import { motion } from 'framer-motion';
+import { Heading } from '@razorpay/blade/components';
+
+const CardHeading = motion(Heading);
+
+<CardHeading layoutId="card-heading" transition={{ duration: theme.motion.duration.slow }}>Hello, World!</CardHeading>
+<CardHeading as="h1" layoutId="card-heading">Hello, World!</CardHeading>
+```
+
+**Cons:**
+
+- We won't be able to preset the animation styles and durations with this approach
+- Can lead to inconsistent animations
+- Inconsistent with other motion presets so not very intuitive and requires learning framer-motion for syntax
+
+#### 3. Exposing `morph` Function from Blade
+
+```jsx
+import { Heading, morph } from '@razorpay/blade/components';
+
+const CardHeading = morph(Heading);
+
+<CardHeading layoutId="card-heading">Hello, World!</CardHeading>
+<Card>
+  <CardBody>
+    <CardHeading as="h1" layoutId="card-heading">Hello, World!</CardHeading>
+    <Box>
+      <Text>Other Text</Text>
+    </Box>
+  </CardBody>
+</Card>
+```
+
+**Pros:**
+
+- Allows presetting animation properties on blade
+
+**Cons:**
+
+- Inconsistent with other motion presets so not very intuitive
+- Comparitively more verbose than suggested API
+
+</details>
+
+#### AnimateInteractions
+
+When we wrap a certain component in AnimateInteractions wrapper from blade, we can animate the children component on interactions of the parent component.
+
+Example in below component, we animate div's background colors on hover and tap.
+
+```jsx
+import { motion } from 'framer-motion';
+import { AnimateInteractions, Scale, Box } from '@razorpay/blade/components';
+
+<AnimateInteractions>
+  <Card>
+    <CardBody>
+      <motion.div
+        variants={{
+          initial: { backgroundColor: 'white' },
+          hover: { backgroundColor: 'tomato' },
+          tap: { backgroundColor: 'skyblue' },
+        }}
+      >
+        <Text>The box will change colors based on the interactions defined in variants</Text>
+      </motion.div>
+    </CardBody>
+  </Card>
+</AnimateInteractions>;
+```
+
+In most cases, we have interaction presets defined that will help you animate things without explicitly defining variants.
+
+#### Scale
+
+Scale animation can be used indepedently to scale item on certain actions but also inside AnimateInteractions.
+
+<table>
+<tr>
+  <td width="700px">
+In below example, the images scales up when its parent container is hovered<br/><br/>
+
+<!-- prettier-ignore -->
+```jsx
+import { 
+  AnimateInteractions, 
+  Scale 
+} from '@razorpay/blade/components';
+
+<AnimateInteractions>
+  <Card>
+    <CardBody>
+      <Scale>
+        <img src="./rajorpay.jpeg" />
+      </Scale>
```

Comment:
This feels a bit "Magic" with AnimateInteraction automatically making the Scale working. 

What if I want the Scale to trigger based upon some other state or timer? 

Eg: Scale an icon button to make attract attention after a set duration.

Reply:
And in here what is AnimateInteraction doing? It will create another wrapper with hover events attached? 

Define behaviour of:
1. What happens on focusing the AnimateInteractions? should it have focus? 
2. How will it work in conjunction with shouldScaleOnHover prop of Card
3. AnimateInteractions can be inside Card Body?

Reply:
> This feels a bit "Magic" with AnimateInteraction automatically making the Scale working.

True true. I didn't know how the props would be defined while writing this. I feel this feels less magic (this is the final code after implementation) -

### Scale when parent is hovered
```jsx
<AnimateInteraction motionTriggers={['hover']}>
  <SomeOtherCompoennt>
    <Scale><ComponentToScale /></Scale>
  </SomeOtherComponents>
</AnimateInteraction>
```

### Scale when parent is tapped
```jsx
<AnimateInteraction motionTriggers={['tap']}>
  <SomeOtherCompoennt>
    <Scale><ComponentToScale /></Scale>
  </SomeOtherComponents>
</AnimateInteraction>
```


### Scale when the component itself is hovered
```jsx
<Scale motionTriggers={['hover']}><ComponentToScale /></Scale>
```